philipmeneghini/InstaConnect
# Realtime Notification Service

**Summary**
This PR creates a new notification service in the backend which is utilized to push and read notifications for different users. Currently notifications are only pushed for new likes and followers. In order to have these notifications update realtime this new service utilizes the `SignalR` package to create a websocket connection between the frontend and backend. This enables for new notifications to come in without any refreshing of the window in the client's current session.

**NotificationService**

- `GET`, `PUT`, `DELETE` endpoints available at `/Notification` or `/Notifications`
- Notification Hub utilizes `/NotificationHub` to connect to the client and receive realtime updates with `SignalR`
- Uses the new `notifications` folder in the Mongo database to store older notifications

**NotificationProvider**

- A `REACT` Context which is used to provide these notifications across the application using the `useContext` hook
- Handles the connection to the Websocket for the entirety of the frontend application
- retrieves older notifications from the database upon startup
![RealtimeNotificaitons](https://github.com/philipmeneghini/InstaConnect/assets/63440066/2b116409-d20d-4322-8aaa-a5599acd1cec)

**Testing**

- Make sure all the new endpoints work as expected
- Test that changes in the frontend get saved in the database's new `notifications` section
- Verify that new notifications are pushed automatically without any browser refresh
